File: ObscuritasMediaManager.Frontend/data/filter-entry.js
import { CheckboxState } from './enumerations/checkbox-state.js';

/**
 * @template U
 * @template {Object.<string, U> | U[]} T
 */
export class FilterEntry {
    get required() {
        return Object.entries(this.states)
            .filter((x) => x[1] == CheckboxState.Allow)
            .map((x) => /** @type {U} */ (x[0]));
    }

    get ignored() {
        return Object.entries(this.states)
            .filter((x) => x[1] == CheckboxState.Ignore)
            .map((x) => /** @type {U} */ (x[0]));
    }

    get forbidden() {
        return Object.entries(this.states)
            .filter((x) => x[1] == CheckboxState.Forbid)
            .map((x) => /** @type {U} */ (x[0]));
    }

    /**@type  {Object.<U, CheckboxState>} */ states;

    /**
     * @param {T} type
     * @param {CheckboxState} defaultValue
     */
    constructor(type, defaultValue = CheckboxState.Ignore) {
        this.states = /** @type {any} */ ({});
        this._type = Object.values(type)[0];
        Object.values(type).forEach((x) => (this.states[x] = defaultValue));
    }
}

File: ObscuritasMediaManager.Client/ObscuritasMediaManager.Client/Data/FilterEntry.cs
using System;
using System.Linq;

namespace ObscuritasMediaManager.Client.Data;

public class FilterEntry<T> where T : notnull
{
    public Dictionary<T, CheckboxState>   states = new();

    public IEnumerable<T> forbidden => states.Where((x) => x.Value == CheckboxState.Forbid).Select((x) => x.Key);

    public IEnumerable<T> ignored => states.Where((x) => x.Value == CheckboxState.Ignore).Select((x) => x.Key);

    public IEnumerable<T> required => states.Where((x) => x.Value == CheckboxState.Require).Select((x) => x.Key);

    public FilterEntry(IEnumerable<T> values, CheckboxState defaultValue = CheckboxState.Ignore)
    {
        states = values.ToDictionary(x => x, x => defaultValue);
    }

    public void ChangeAll(CheckboxState target)
    {
        foreach (var state in states) states[state.Key] = target;
    }
}
